heat/core/manipulations.py

```diff
@@ -2779,146 +2903,96 @@ def unique(
-            a.comm.Allgatherv(
-                inverse_pos, (inverse_buf, inverse_counts, inverse_displs), recv_axis=0
-            )
-            inverse_buf = inverse_buf.transpose(0, a.split)
-
-        # Run unique a second time
-        gres = torch.unique(gres_buf, sorted=sorted, return_inverse=return_inverse, dim=unique_axis)
-        if return_inverse:
-            # Use the previously gathered information to generate global inverse_indices
-            g_inverse = gres[1]
-            gres = gres[0]
-            if axis is None:
-                # Calculate how many elements we have in each layer along the split axis
-                elements_per_layer = 1
-                for num, val in enumerate(a.gshape):
-                    if not num == a.split:
-                        elements_per_layer *= val
-
-                # Create the displacements for the flattened inverse indices array
-                local_elements = [displ * elements_per_layer for displ in inverse_displs][1:] + [
-                    float("inf")
-                ]
-
-                # Flatten the inverse indices array every element can be updated to represent a global index
-                transposed = inverse_buf.transpose(0, a.split)
-                transposed_shape = transposed.shape
-                flatten_inverse = transposed.flatten()
-
-                # Update the index elements iteratively
-                cur_displ = 0
-                inverse_indices = [0] * len(flatten_inverse)
-                for num in range(len(inverse_indices)):
-                    if num >= local_elements[cur_displ]:
-                        cur_displ += 1
-                    index = flatten_inverse[num] + displs[cur_displ]
-                    inverse_indices[num] = g_inverse[index].tolist()
-
-                # Convert the flattened array back to the correct global shape of a
-                inverse_indices = torch.tensor(inverse_indices).reshape(transposed_shape)
-                inverse_indices = inverse_indices.transpose(0, a.split)
-
-            else:
-                inverse_indices = torch.zeros_like(inverse_buf)
-                steps = displs + [None]
-
-                # Algorithm that creates the correct list for the reverse_indices
-                for i in range(len(steps) - 1):
-                    begin = steps[i]
-                    end = steps[i + 1]
-                    for num, x in enumerate(inverse_buf[begin:end]):
-                        inverse_indices[begin + num] = g_inverse[begin + x]
-
+        lres = torch.unique(local_data, sorted=True, return_inverse=False, dim=unique_axis)
+    gres = factories.array(lres, dtype=a.dtype, is_split=0, device=a.device)
+
+    # calculate size (bytes) of local unique. If less than local_data, gather and run everything locally
+    _, data_max_lshape, _ = a.comm.chunk(a.gshape, a.split, rank=0)
+    data_max_lbytes = torch.prod(torch.tensor(data_max_lshape)) * a.larray.element_size()
```

Comment:
dont need to put everything in tensors here (im not sure about the overhead) but this can also be done with the max `lshape * element_size` i believe that `element_size` returns an int but i am not sure

Reply:
modified, now using `torch.prod(a.lshape[0])* a.larray.element_size()`